src/bun.js/node/hdr_histogram.zig

```diff
@@ -0,0 +1,503 @@
+        if (self.max < value) self.max = value;
+    }
+
+    fn calculate_index(self: *const This, value: u64) usize {
+        const bucket_index = self.get_bucket_index(value);
+        const sub_bucket_index = self.get_sub_bucket_index(value, bucket_index);
+        const counts_index = self.get_counts_index(bucket_index, sub_bucket_index);
+        return counts_index;
+    }
+
+    fn get_counts_index(self: *const This, bucket_index: u64, sub_bucket_index: u64) usize {
+        const bucket_base_index = (bucket_index + 1) << self.sub_bucket_half_count_magnitude;
+        return @as(usize, bucket_base_index + sub_bucket_index - self.sub_bucket_half_count);
+    }
+
+    fn get_bucket_index(self: *const This, value: u64) u8 {
+        const pow2ceiling = 64 - @clz(value | self.sub_bucket_mask);
+        return pow2ceiling - self.unit_magnitude - (self.sub_bucket_half_count_magnitude + 1);
+    }
+
+    fn get_sub_bucket_index(self: *const This, value: u64, bucket_index: u8) u64 {
+        return value >> @as(u6, @intCast(bucket_index + self.unit_magnitude));
+    }
+
+    //
+    // Reads from the histogram
+    //
+
+    fn value_from_index(self: *This, index: u64) u64 {
+        const bucket_index = self.get_bucket_index_from_idx(index);
+        const sub_bucket_index = self.get_sub_bucket_index_from_idx(index, bucket_index);
+
+        // Directly compute the value from the bucket index and sub-bucket index
+        return @as(u64, sub_bucket_index) << @as(u6, @truncate(bucket_index + self.unit_magnitude));
+    }
+
+    fn get_bucket_index_from_idx(self: *const This, index: u64) u8 {
+        var bucket_index: u8 = 0;
+        var remaining_index = index;
+
+        while (remaining_index >= self.sub_bucket_count) {
+            bucket_index += 1;
+            remaining_index -= self.sub_bucket_half_count;
+        }
+        return bucket_index;
+    }
+
+    fn get_sub_bucket_index_from_idx(self: *const This, index: u64, bucket_index: u8) u64 {
+        const sub_bucket_index = index - (bucket_index * self.sub_bucket_half_count);
+        return sub_bucket_index;
+    }
+
+    // percentile is a value between 0 and 100
+    pub fn value_at_percentile(self: *This, percentile: f64) ?u64 {
+        if (percentile < 0 or percentile > 100) {
+            return null;
+        }
+        if (percentile == 0 and self.total_count > 0) {
+            return self.min;
+        }
+        const total = self.total_count;
+        const target_count = @as(f64, @floatFromInt(total)) * percentile / 100;
+        var running_total: f64 = 0;
+        for (self.counts, 0..) |count, index| {
+            running_total += @floatFromInt(count);
+            if (running_total >= target_count) {
+                // we found the index that corresponds to the percentile
+                return self.value_from_index(index);
+            }
+        }
+        return null;
+    }
+};
+
+test "value_at_percentile" {
```

Comment:
i enjoy seeing thorough testing.

i wish we had a mechanism for running Zig based tests in our codebase.

Reply:
TBH did this because unit testing zig test `just_my_file.zig` is so much faster than doing a full build and integration test

And I'm happy to. Tests good.